lib/sidekiq/processor.rb

```diff
@@ -10,36 +8,65 @@ module Sidekiq
   # processes it.  It instantiates the worker, runs the middleware
   # chain and then calls Sidekiq::Worker#perform.
   class Processor
-    # To prevent a memory leak, ensure that stats expire. However, they should take up a minimal amount of storage
-    # so keep them around for a long time
+
+    # To prevent a memory leak, ensure that stats expire. However, they
+    # should take up a minimal amount of storage so keep them around
+    # for a long time.
     STATS_TIMEOUT = 24 * 60 * 60 * 365 * 5
 
     include Util
-    include Actor
-
-    def self.default_middleware
-      Middleware::Chain.new do |m|
-        m.add Middleware::Server::Logging
-        m.add Middleware::Server::RetryJobs
-        if defined?(::ActiveRecord::Base)
-          require 'sidekiq/middleware/server/active_record'
-          m.add Sidekiq::Middleware::Server::ActiveRecord
-        end
-      end
+
+    attr_reader :thread
+
+    def initialize(mgr)
+      @mgr = mgr
+      @done = false
+      @work = ::Queue.new
     end
 
-    attr_accessor :proxy_id
+    def terminate(wait=false)
+      @done = true
+      @work << nil
+      @thread.value if wait
+    end
 
-    def initialize(boss)
-      @boss = boss
+    def kill(wait=false)
+      # unlike the other actors, terminate does not wait
+      # for the thread to finish because we don't know how
+      # long the job will take to finish.  Instead we
+      # provide a `kill` method to call after the shutdown
+      # timeout passes.
+      @thread.raise ::Sidekiq::Shutdown
```

Comment:
I've read in a number of places that `Thread#raise` and `Thread#kill` are unsafe.  Has that changed?